### PR TITLE
feat(UI): Attempt on displaying an error when sourcing wrong files.

### LINKED ISSUE #49492: Show warning to users when they try to source the wrong files
### Is your feature request related to a problem? Please describe.

When a user tries to source `style.css` (rather than `styles.css`) this fails silently.  Similarly, sourcing anything other than `script.js` will result in a rather cryptic error: `Uncaught SyntaxError: Unexpected token '<'`.

This is needlessly frustrating.

### Describe the solution you'd like

The console should show a helpful warning. For example

> You have tried to source style.css, but that does not exist. The only files that can be sourced are styles.css and script.js.

### Describe alternatives you've considered

Nothing comes to mind.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by CallmeHongmaybe**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #49492

### CODE REVIEW

**Comment by CallmeHongmaybe** [comment]
@ojeytonwilliams The issue discussed earlier in contributor chat is resolved. But one issue remains is that how I should source the html of the preview in the challenge saga. 
Is it ok to just get the iframe from the document `document.querySelector("iframe")` and just look for any problematic links and scripts inside, or do you need a more complicated solution ?

**Comment by ojeytonwilliams** [comment]
Good question, @CallmeHongmaybe. I think the appropriate place for this code is https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/templates/Challenges/utils/frame.ts since that's where we keep the code that interacts with the preview and test iframes. `initMainFrame` should call that code, so learners will see the message whenever their code has a problem.

One other thing is that it shouldn't throw the error - we just want to warn people that something is wrong. Instead of throwing, could you send the message to the console and the proxyLogger?

**Comment by CallmeHongmaybe** [comment]
image

I'm able to bring the message to the console and the proxyLogger, but one thing that bugged me is that the response always returns 200 no matter what URL I throw in to the `fetch` function.

File: client/src/templates/Challenges/utils/frame.ts [hunk 1]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+        let errMsgs = '';
+        let unfetchedLinks = allLinks.length;
+
+        if (!unfetchedLinks) return null;
+
+        for (const url of allLinks) {
+          const urlAddr =
+            url.tagName === 'SCRIPT'
+              ? (url as HTMLScriptElement).src
+              : (url as HTMLLinkElement).href;
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:292]:
```suggestion
          const url =
            elem.tagName === 'SCRIPT'
              ? (elem as HTMLScriptElement).src
              : (elem as HTMLLinkElement).href;
```

File: client/src/templates/Challenges/utils/frame.ts [hunk 2]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+
+          try {
+            const response = await fetch(String(urlAddr));
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:295]:
```suggestion
            const response = await fetch(url);
```
It's already a string.

File: client/src/templates/Challenges/utils/frame.ts [hunk 3]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+            if (!response.ok) {
+              errMsgs = `${errMsgs}\nCannot retrieve ${urlAddr}, link status code: ${response.status}`;
+            } else if (!unfetchedLinks) {
+              continue;
+            }
+            unfetchedLinks -= 1;
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:301]:
```suggestion
            return response.ok
              ? ''
              : `Cannot retrieve ${url}, link status code: ${response.status}`;
```
What was the logic behind updating `unfetchedLinks`? It seems to me that either all the links can be fetched (and we're fine) or not (and we should show an error). I couldn't see how `unfetchedLinks` fit into that.

File: client/src/templates/Challenges/utils/frame.ts [hunk 4]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+          } catch {
+            errMsgs = `${errMsgs}\n${urlAddr} does not exist. Only files that can be sourced are styles.css, script.js, or remote files`;
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:303]:
```suggestion
            // TODO: tailor message to the file (e.g. script.js, styles.css etc.)
            return `${url} does not exist. Only files that can be sourced are styles.css, script.js, or remote files`;
```
The message is fine for this PR, but we should address the TODO at some piont!

File: client/src/templates/Challenges/utils/frame.ts [hunk 5]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+          }
+        }
+
+        return errMsgs.length ? errMsgs : 'This link does not exist.';
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:307]:
```suggestion
        return (await Promise.all(errors)).filter(Boolean).join('\n');
```

File: client/src/templates/Challenges/utils/frame.ts [hunk 6]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+      })
+      .then(importResult => {
+        if (importResult) {
+          console.warn.call(frameContext.window, importResult);
+          proxyLogger && proxyLogger(importResult);
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:312]:
```suggestion
          // @ts-expect-error the window types will be updated in another PR
          frameContext.window.console.error(importResult);
```
Unfortunately the types are wrong, but I'm already working on a PR that fixes them, so no need to worry about that here!

Anyways, we can make use of the fact that the frame's console already handles the proxying.

File: client/src/templates/Challenges/utils/frame.ts [hunk 7]
```diff
@@ -301,6 +337,9 @@ function handleDocumentNotFound(err: string) {
   if (err !== DOCUMENT_NOT_FOUND_ERROR) {
     console.log(err);
   }
+  // else if (err == FILE_NOT_FOUND_ERROR) {
+  //   console.log(err);
+  // }
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/utils/frame.ts:342]:
```suggestion
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @CallmeHongmaybe, thanks again for putting this together.

Your code's functionality was great. Mostly I found stylistic issues and they cascaded into a lot of comments.

**Comment by ojeytonwilliams** [comment]
> But one thing that bugged me is that the response always returns 200 no matter what URL I throw in to the fetch function.

Sorry, missed this comment :man_facepalming: 

So, yeah, relative URLs will all be 200s in development because Gatsby handles requests for missing pages and returns a 200 status. Requests to other sites should behave correctly, though.

File: client/src/templates/Challenges/utils/frame.ts (hunk 3 above)
Comment (CallmeHongmaybe) [reply client/src/templates/Challenges/utils/frame.ts:301]: The idea behind `unfetchedLinks` is that I want to gather all error messages from multiple fetching errors before displaying it in the `proxyLogger`. But I think it should be ok to get rid of it.

File: client/src/templates/Challenges/utils/frame.ts [hunk 8]
```diff
@@ -273,11 +273,47 @@ const initTestFrame = (frameReady?: () => void) => (frameContext: Context) => {
...
+        const iframeBuildDoc = parser.parseFromString(iframeBuild, 'text/html');
+        const allLinks = iframeBuildDoc.querySelectorAll(
+          'script[src], link[href]'
+        );
+        let errMsgs = '';
+        let unfetchedLinks = allLinks.length;
+
+        if (!unfetchedLinks) return null;
+
+        for (const url of allLinks) {
```
[the remark this answers is not in the document]
Comment (CallmeHongmaybe) [reply client/src/templates/Challenges/utils/frame.ts:288]: There's already an `async` in line 276.

File: client/src/templates/Challenges/utils/frame.ts (hunk 4 above)
Comment (CallmeHongmaybe) [reply client/src/templates/Challenges/utils/frame.ts:303]: What's the usual deadline for TODOs like this ?

File: client/src/templates/Challenges/utils/frame.ts (hunk 6 above)
Comment (CallmeHongmaybe) [reply client/src/templates/Challenges/utils/frame.ts:312]: Thanks for letting me know.

File: client/src/templates/Challenges/utils/frame.ts (hunk 7 above)
Comment (CallmeHongmaybe) [reply client/src/templates/Challenges/utils/frame.ts:342]: I see it now. Thanks.

File: client/src/templates/Challenges/utils/frame.ts (hunk 5 above)
Comment (CallmeHongmaybe) [reply client/src/templates/Challenges/utils/frame.ts:307]: May I ask why you would want to filter out `null` or `undefined` values when `errors` would be an array of all strings?

**Comment by CallmeHongmaybe** [comment]
I tried with absolute URLs like `https://fonts.googleapis.com/css?family=Lobster` and jQuery CDNs and regardless of the spelling the `errors` return a fulfilled promise.
Are there any environments in FCC ( like preview or staging kind of environment ) to test the fetching ?

File: client/src/templates/Challenges/utils/frame.ts (hunk 4 above)
Comment (naomi-lgbt) [reply client/src/templates/Challenges/utils/frame.ts:303]: Ideally it would be something we address in this PR.

**Comment by CallmeHongmaybe** [comment]
Hello is anyone here ?

File: client/src/templates/Challenges/utils/frame.ts (hunk 5 above)
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/utils/frame.ts:307]: This filters out falsy values and so removes the empty strings.

**Comment by ojeytonwilliams** [comment]
Hi @CallmeHongmaybe, the review suggestions still need to be addressed before we can move forwards.

As for testing, you can check the behaviour for external urls that you know return 404 statuses and/or build the site locally. While Gatsby's development mode won't work, the site it builds will behave properly.

**Pull request closed without merging**